core/field.js

```diff
@@ -415,6 +415,32 @@ Blockly.Field.prototype.toXml = function(fieldElement) {
   return fieldElement;
 };
 
+/**
+ * Saves this fields value as something which can be serialized to JSON. Should
+ * only be called by the serialization system.
+ * Default implementation calls the XML versions for backwards compatibility.
+ * @return {*} JSON serializable state.
+ * @package
+ */
+Blockly.Field.prototype.saveState = function() {
+  // Default backwards-compatible implementation.
+  var container = Blockly.utils.xml.createElement('field');
+  container.setAttribute('name', this.name || '');
+  return Blockly.Xml.domToText(this.toXml(container));
+};
```

Comment:
Note that this doesn't solve for the case where someone is extending a field other than Field and adding custom serialization. For example, pretend `FieldVariable` is external so new hooks aren't being added to it. `FieldVariable` extends `FieldDropdown` and adds custom serialization with `toXml`. When we go to serialize it using `saveState` `FieldDropdown`'s `saveState` function is hit instead of this one on `Fiield`. Therefore `FieldVariable` gets serialized as a dropdown field rather than stringifying its XML implementation.

Sadly I don't think this is solvable :/ We can't do any fancy undefined checking like we do with blocks without requiring devs to explicitly implement serialization on every field (even if they want to use the default implementation).

This is why I hate allowing people to depend on the internal class hierarchy lol.

Reply:
Hmm so what does this mean for external developers? Any custom field that adds logic to the `toXml` method will need to add a `saveState` method?

Reply:
Only custom fields with custom `toXml` method that extend classes other than `Blockly.Field`. So if you extend `Blockly.TextInput`, `Blockly.FieldDropdown` etc, then yeah you have to add a `saveState` method :/

I think it may be better to rip the backwards compatibility out for now, and explore providing a python script that will add wrapper implementations to developers' block and field files. That will make make it so that our internal code doesn't rely on `Blockly.Xml` (making it possible to delete later) and it fixes the above problem. I also mention that option [here](https://github.com/google/blockly/pull/5054#discussion_r672660463).

Reply:
Ok, that sounds fine with me. If we run into any unknown problems with the script we can add this back before releasing.